### PR TITLE
[Merged by Bors] - feat: port Computability.TuringMachine

### PR DESCRIPTION
**Pull request by int-y1**
---

This file uses `local notation` to make the types more concise. (Very few other files used `local notation` in this way.)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by int-y1** [comment]
* The first half of the file is pretty much done.
* I'm giving up because I ran into `don't know how to synthesize implicit argument` a lot of times, and I couldn't fix those.
* I recommend doing these comments first: `Porting note: TODO clean up this ugly hack`

**Comment by int-y1** [comment]
Remaining tasks:

* [ ] Fix `wlog` in `Turing.ListBlank.ext`
* [ ] Fix error with `suffices` in `Turing.TM1to1.stepAux_read`
* [ ] Lots of errors in `namespace TM2to1`

File: Mathlib/Computability/TuringMachine.lean [hunk 1]
```diff
@@ -0,0 +1,2859 @@
...
+  | TM1.Stmt.move d q, v => ((some q, v), move d)
+  | TM1.Stmt.write a q, v => ((some q, v), write (a s v))
+  | TM1.Stmt.load a q, v => trAux s q (a s v)
+  | TM1.Stmt.branch p q₁ q₂, v => cond (p s v) (trAux s q₁ v) (trAux s q₂ v)
+  | TM1.Stmt.goto l, v => ((some (M (l s v)), v), write s)
+  | TM1.Stmt.halt, v => ((none, v), write s)
+#align turing.TM1to0.tr_aux Turing.TM1to0.trAux
+
+-- mathport name: exprcfg₀
+local notation "Cfg₁₀" => TM0.Cfg Γ Λ'₁₀
```
Comment (dupuisf) [inline Mathlib/Computability/TuringMachine.lean:1501]:
```suggestion
local notation "Cfg₁₀" => TM0.Cfg Γ Λ'₁₀
```

File: Mathlib/Computability/TuringMachine.lean [hunk 2]
```diff
@@ -0,0 +1,2859 @@
...
+
+variable {Λ : Type _} [Inhabited Λ]
+
+variable {σ : Type _} [Inhabited σ]
+
+-- mathport name: exprstmt₁
+local notation "Stmt₁" => Stmt Γ Λ σ
+
+-- mathport name: exprcfg₁
+local notation "Cfg₁" => Cfg Γ Λ σ
```
Comment (dupuisf) [inline Mathlib/Computability/TuringMachine.lean:1657]:
```suggestion
local notation "Stmt₁" => Stmt Γ Λ σ

local notation "Cfg₁" => Cfg Γ Λ σ
```

File: Mathlib/Computability/TuringMachine.lean [hunk 3]
```diff
@@ -0,0 +1,2859 @@
...
+local notation "Λ'₁" => @Λ' Γ Λ σ  -- Porting note: Added this to clean up types.
+
+instance : Inhabited Λ'₁ :=
+  ⟨Λ'.normal default⟩
+
+-- mathport name: exprstmt'
+local notation "Stmt'₁" => Stmt Bool Λ'₁ σ
+
+-- mathport name: exprcfg'
+local notation "Cfg'₁" => Cfg Bool Λ'₁ σ
```
Comment (dupuisf) [inline Mathlib/Computability/TuringMachine.lean:1674]:
```suggestion
local notation "Stmt'₁" => Stmt Bool Λ'₁ σ

local notation "Cfg'₁" => Cfg Bool Λ'₁ σ
```

File: Mathlib/Computability/TuringMachine.lean [hunk 4]
```diff
@@ -0,0 +1,2859 @@
...
+#align turing.TM1to1.supports_stmt_read Turing.TM1to1.supportsStmt_read
+
+variable (enc0 : enc default = Vector.replicate n false)
+
+section
+
+variable {enc}
+
+--include enc0  -- Porting note: `include` doesn't exist
+
```
Comment (dupuisf) [inline Mathlib/Computability/TuringMachine.lean:1755]:
```suggestion

```
These `include` and `omit` statements can just be deleted outright.

File: Mathlib/Computability/TuringMachine.lean [hunk 5]
```diff
@@ -0,0 +1,2859 @@
...
+/-- The machine configuration translation. -/
+def trCfg : Cfg₁ → Cfg'₁
+  | ⟨l, v, T⟩ => ⟨l.map Λ'.normal, v, trTape enc0 T⟩
+#align turing.TM1to1.tr_cfg Turing.TM1to1.trCfg
+
+variable {enc}
+
+--include enc0  -- Porting note: `include` doesn't exist
+
```
Comment (dupuisf) [inline Mathlib/Computability/TuringMachine.lean:1791]:
```suggestion

```

File: Mathlib/Computability/TuringMachine.lean [hunk 6]
```diff
@@ -0,0 +1,2859 @@
...
+  cases' l₂' with b l₂' <;> simp only [List.length_nil, List.length_cons, Nat.succ_inj'] at e
+  rw [List.reverseAux, ← IH (a :: l₁) l₂' e]
+  simp only [stepAux, ListBlank.append, Tape.write_mk', Tape.move_right_mk', ListBlank.head_cons,
+    ListBlank.tail_cons]
+#align turing.TM1to1.step_aux_write Turing.TM1to1.stepAux_write
+
+variable (encdec : ∀ a, dec (enc a) = a)
+
+--include encdec  -- Porting note: `include` doesn't exist
+
```
Comment (dupuisf) [inline Mathlib/Computability/TuringMachine.lean:1844]:
```suggestion

```

**Review by dupuisf — COMMENTED** [review]
Thanks! I pretty much only have cosmetic remarks.

bors d+

**Comment by int-y1** [comment]
:tada:
bors r+

**Pull request closed without merging**